recipes-omnect/initrdscripts/omnect-os-initramfs/factory-reset

```diff
@@ -216,70 +243,78 @@ factory_reset_run() {
         return 1
     fi
 
-    factory_reset_mode=$(get_bootloader_env_var "${FACTORY_RESET_VAR}")
-    if [ -z "${factory_reset_mode}" ]; then
-        return  # no factory reset triggered; normal boot
+    local mode=$(echo "${factory_reset}" | jq -e -r .mode) || { handle_status "1" "-" "mode_not_set" ""; return 0; }
+    # don't error if preserve is not set
+    local preserve=$(echo "${factory_reset}" | jq -r '.preserve')
+    if [[ "null" != "${preserve}" ]]; then
+        preserve=($(echo "${factory_reset}" | jq -e -r '.preserve | .[]')) || { handle_status "3" "-" "could_not_parse_preserve_array" ""; return 0; }
+    else
+        preserve=()
     fi
 
-    factory_reset_restore_list=$(get_bootloader_env_var "${FACTORY_RESET_RESTORE_LIST_VAR}")
-
-    # we already mounted /dev/omnect/boot when we're using grub. so we unmount so we don't disturb the restore mount/umount handling
-    if [ -f /usr/bin/grub-editenv ]; then
-        sync
-        run_cmd umount ${ROOTFS_DIR}/boot
+    msg "Running factory-reset [mode=${mode}, preserve=${preserve[@]}]..."
+    factory_reset_mount
+
+    # create preserve_list
+    declare -a preserve_list
+    preserve_list+=("/etc/omnect/factory-reset.d/")
+    if [[ " ${preserve[@]} " =~ " applications " ]]; then
+        preserve=( ${preserve[@]/applications} )
+        for i in $(find ${ROOTFS_DIR}/etc/omnect/factory-reset.d/ -iname *.json); do
+            preserve_list+=( $(cat ${i} | jq -r -e '.paths[]') )
+            preserve_list_err=$?
+            [[ 0 -ne ${preserve_list_err} ]] && { handle_status "3" "${preserve_list_err}" "${i/${ROOTFS_DIR}}:paths" "${preserve_list}"; factory_reset_umount; return 0; }
+        done
     fi
 
-    msg "Running Factory Reset [mode=${factory_reset_mode}]..."
-
-    if [ -n "${factory_reset_restore_list}" ]; then
-        msg "restore list present"
-        factory_reset_mount  # logical filesystem tree
+    for key in ${preserve[@]}; do
+        preserve_list+=( $(cat ${ROOTFS_DIR}/etc/omnect/factory-reset.json | jq -r -e --arg key "${key}" .$key[]) )
```

Comment:
again stdin redirection could be used.
and maybe it is better to surround `${preserve[@]}` with double quotes just for the case of strange filenames.

Reply:
> maybe it is better to surround ${preserve[@]} with double quotes just for the case of strange filenames.

please elaborate. if these files are added to the array correctly, this shouldn't be a problem.
can you give an example which would fail?

Reply:
parameter expansion of `${preserve[@]}` underlies standard word splitting in the for loop construct, weird elements of `preserve` array containing spaces would lead ti errors.
using `for key in "${preserve[@]}" ...` would prevent that.

Reply:
an example would be easier to understand what you want from me.
e.g. unexpected but accepted:
```
$ foo=( "1" "2" "3" )
$ foo+=( "4 5" )
$ for i in ${foo[@]}; do echo $i; done
1
2
3
4
5
$ for i in "${foo[@]}"; do echo $i; done
1
2
3
4 5

```

Reply:
Will be handled in a future PR.
I'll will add a test which contains filenames with Spaces (and other not so nice Tokens - Ideas welcome),